Add test for schema evolution of auto_ptr -> unique_ptr

makortel opened the pull request:
#### PR description:

As a stepping stone to solve https://github.com/cms-sw/cmssw/issues/43422 and https://github.com/cms-sw/cmssw/issues/43923, this PR extends the schema evolution test to cover a case where an `std::auto_ptr` class member is evolved to `std::unique_ptr`. With the first commit only and ROOT older than 6.36 the test will fail.

The second commit applies the recipe https://github.com/cms-sw/cmssw/issues/43923#issuecomment-2898938773 to the test.

Needs the test file update from https://github.com/cms-data/IOPool-Input/pull/4

Resolves https://github.com/cms-sw/framework-team/issues/1451

#### PR validation:

In ROOT6.36 the schema evolution unit test succeeds.

makortel committed:
Add test for schema evolution of auto_ptr -> unique_ptr

makortel committed:
Add read rules for auto_ptr->unique_ptr using ROOT 6.36 features

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48817/45934

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  IOPool/Input/test/SchemaEvolutionTestRead.cc  modified in PR(s): #44372
  - File  IOPool/Input/test/SchemaEvolutionTestWrite.cc  modified in PR(s): #44372

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48817/45934/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48817/45934/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

makortel commented:
@cmsbuild, please test with https://github.com/cms-data/IOPool-Input/pull/4

This test should fail, but I want it on the record.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 40KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c50cc/47900/summary.html
**COMMIT**: 15f2fea5255d9ead0908110069b4c3e6d55e4620
**CMSSW**: CMSSW_15_1_X_2025-08-28-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48817/47900/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @belforte cms-sw/cmsdist#10047

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c50cc/47900/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c50cc/47900/git-merge-result


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestIOPoolInputSchemaEvolution had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [added 2 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-28-1100+0c50cc/70040/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-28-1100+0c50cc/70040/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4115479
 - DQMHistoTests: [Total failures: 21](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-28-1100+0c50cc/70040/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4115438
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-28-1100+0c50cc/70040/triggerResults): no differences found

makortel commented:
test parameters:
  - pull_request = https://github.com/cms-data/IOPool-Input/pull/4
  - release = CMSSW_15_1_ROOT636_X

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48817/45941

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  IOPool/Input/test/SchemaEvolutionTestRead.cc  modified in PR(s): #44372
  - File  IOPool/Input/test/SchemaEvolutionTestWrite.cc  modified in PR(s): #44372

makortel commented:
@cmsbuild, please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c50cc/47910/summary.html
**COMMIT**: 317c8dbe35038744117c5c7049d93864baa0cf68
**CMSSW**: CMSSW_15_1_ROOT636_X_2025-08-27-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48817/47910/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @CMS-HGCAL cms-sw/cmssw#48812
 - @bsunanda cms-sw/cmssw#48800
 - @Parsifal-2045 cms-sw/cmssw#48774
 - @makortel cms-sw/cmssw#48783
 - @cms-sw cms-sw/cmsdist#10045
 - @belforte cms-sw/cmsdist#10047

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c50cc/47910/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0c50cc/47910/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially [added 187 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-27-2300+0c50cc/70050/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-27-2300+0c50cc/70050/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4115479
 - DQMHistoTests: [Total failures: 579](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-27-2300+0c50cc/70050/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4114880
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 262.5 KiB( 49 files compared)
 - DQMHistoSizes: changed ( 1000.0,... ): 6.250 KiB        DQM/TimerService
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-27-2300+0c50cc/70050/triggerResults): no differences found

pcanal commented on DataFormats/TestObjects/src/classes_def.xml:
```diff
@@ -350,70 +369,83 @@ exception when running testMissingDictionaryChecking_cfg.py.
   <version ClassVersion="4" checksum="3437951039"/>
+  <version ClassVersion="3" checksum="1545257825"/>
  </class>
+ <class name="std::auto_ptr<edmtest::SchemaEvolutionContained>"/>
```
The scope is any library/dictionary that is being loaded by the failing job.  Finding out what is the class is question is 'straightforward' by getting the name of the 'this' in the following stack frames:
```
#4  0x00007ffff788c074 in TClass::BuildRealData(void*, bool) () from /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw/CMSSW_15_1_ROOT636_X_2025-08-20-2300/external/el8_amd64_gcc12/lib/libCore.so
#5  0x00007ffff0a37a10 in TStreamerInfo::BuildOld() () from /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw/CMSSW_15_1_ROOT636_X_2025-08-20-2300/external/el8_amd64_gcc12/lib/libRIO.so
#6  0x00007ffff78789a8 in TClass::GetStreamerInfoImpl(int, bool) const () from /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw/CMSSW_15_1_ROOT636_X_2025-08-20-2300/external/el8_amd64_gcc12/lib/libCore.so
#7  0x00007ffff7878c8e in TClass::GetStreamerInfo(int, bool) const () from /cvmfs/cms-ib.cern.ch/sw/x86_64/week1/el8_amd64_gcc12/cms/cmssw/CMSSW_15_1_ROOT636_X_2025-08-20-2300/external/el8_amd64_gcc12/lib/libCore.so
```
At the very least that information will get us closer to understanding what is going on.

makortel replied:
> I expect to see `GetName` being `auto_ptr` at stack # 7

Correct, `this->GetName()` frame 7 is `auto_ptr<edmtest::SchemaEvolutionContained>`

> what is `GetName` at stack # 16? (and the result of `this->ls()`)

```
(gdb) p this
$3 = (TStreamerInfo * const) 0x7fffc72677e0
(gdb) p this->GetName()
$4 = 0x7fffc9da74e0 "edmtest::VectorVectorMiddle"
(gdb) p this->ls("")

StreamerInfo for class: edmtest::VectorVectorMiddle, version=3, checksum=0x2d410aab
  vector<edmtest::VectorVectorElement> middleVector_   offset=  0 type=300 stl=1 ctype=61
   i= 0, middleVector_   type=300, offset=  0, len=1, method=0
$5 = void
```

makortel replied:
Following @pcanal's suggestion I tested what happens if we would not convert the `Inspection for auto_ptr<` message to an exception in `InitRootHandlers`. By effectively ignoring the message, the test succeeds.

In parallel I initiated a ROOT 6.36 debug build in https://github.com/cms-sw/cmsdist/pull/10059 to give a reproducer for @pcanal.

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_16_0_X. Please open a backport if it should also go in to CMSSW_15_1_X.

makortel committed:
Remove the auto_ptr dictionary and demote the resulting message in InitRootHandlers

The demoted message effectively tells that the dictionary for auto_ptr
is missing. If we don't consider the message as an error, ROOT will
auto-parse the necessary header. This auto-parsing was deemed to be
less bad than keeping the dictionary for auto_ptr (with the compiler
warnings etc).

makortel commented:
Rebased to CMSSW_16_0_0_pre3, and (in a separate commit) removed the `auto_ptr` dictionary and demoted the resulting `Inspection for auto_ptr` to INFO (as discussed in https://github.com/cms-sw/cmssw/pull/48817#discussion_r2334702347)

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @sextonkennedy, @ftenchini, @mandrenguyen (and backports should be raised in the release meeting by the corresponding L2)
**Notice** This PR was tested with additional Pull Request(s), please also merge them if necessary: cms-data/IOPool-Input#4
